### PR TITLE
Add WhenAny #28

### PR DESCRIPTION
**Pull request by Ri7ay**

### CODE REVIEW

File: include/yaclib/async/when_any.hpp [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+    return {std::move(future), new container::Counter<AnyCombinator<T>>{std::move(promise)}};
+  }
+
+  void Combine(util::Result<T> result) {
+    if (_done.load(std::memory_order_acquire)) {
+      return;
+    }
+
+    if (_done.exchange(true, std::memory_order_acq_rel)) {
+      return;
```
Comment (kononovk) [inline include/yaclib/async/when_any.hpp:37]: It seems we can change this to one CAS and avoid invalidating the cache lines.

File: test/unit/async/when_any.cpp [hunk 2]
```diff
@@ -0,0 +1,196 @@
...
+  auto any = async::WhenAny(empty.begin(), empty.end());
+
+  EXPECT_TRUE(any.Ready());
+  EXPECT_NO_THROW(std::move(any).Get().Ok());
+}
+
+// TEST(Vector, EmptyInput) {
+//   EmptyInput();
+// }
+
```
Comment (kononovk) [inline test/unit/async/when_any.cpp:128]: delete unused code

File: test/unit/async/when_any.cpp [hunk 3]
```diff
@@ -0,0 +1,196 @@
...
+TEST(Array, MultiThreaded) {
+  MultiThreaded<TestSuite::Array>();
+}
+
+TEST(VoidArray, MultiThreaded) {
+  MultiThreaded<TestSuite::Array, void>();
+}
+
+}  // namespace
```
Comment (kononovk) [inline test/unit/async/when_any.cpp:196]: Add empty line to the end of file, check `.githooks` to do it automatically

File: include/yaclib/async/when_any.hpp [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+
+template <typename T, typename... Fs>
+Future<T> WhenAny(Future<T>&& head, Fs&&... tail) {
+  static_assert((... && util::IsFutureV<Fs>));
+  auto [future, combinator] = detail::AnyCombinator<T>::Make(1);
+  detail::WhenAnyImpl(combinator, std::move(head), std::move(tail)...);
+  return std::move(future);
+}
+
+}  // namespace yaclib::async
```
Comment (kononovk) [inline include/yaclib/async/when_any.hpp:86]: Add empty line to the end of file, check `.githooks` directory to do it automatically

File: include/yaclib/async/when_any.hpp [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+      return;
+    }
+
+    if (_done.exchange(true, std::memory_order_acq_rel)) {
+      return;
+    }
+    std::move(_promise).Set(std::move(result));
+  }
+
+  AnyCombinator(Promise<T> promise) : _done{false}, _promise{std::move(promise)} {
```
Comment (kononovk) [inline include/yaclib/async/when_any.hpp:42]: Make this constructor explicit

File: include/yaclib/async/when_any.hpp [hunk 6]
```diff
@@ -0,0 +1,86 @@
+#pragma once
```
Comment (MBkkt) [inline include/yaclib/async/when_any.hpp:1]: Very useful pattern:
 Run async operation
 Run same operation, but different context, for example other database instance

 WhenAny(f1, f2).Then(...). ...

In this case, we need the first success, but not the first error. I mean any error is suitable for us, but if we return an error, it should be means that all futures have completed with an error.

**Review by MBkkt — CHANGES_REQUESTED** [review]
Very useful pattern:
 Run async operation
 Run same operation, but different context, for example other database instance

 WhenAny(f1, f2).Then(...). ...

In this case, we need the first success, but not the first error. I mean any error is suitable for us, but if we return an error, it should be means that all futures have completed with an error.

File: include/yaclib/async/when_any.hpp [hunk 7]
```diff
@@ -0,0 +1,162 @@
...
+      if (!_done.exchange(true, std::memory_order_acq_rel)) {
+        std::move(_promise).Set(std::move(result));
+      }
+    } else if (_size.fetch_sub(1, std::memory_order_acq_rel) <= 1U) {
+      std::move(_promise).Set(std::move(result));
+    }
+  }
+
+  explicit AnyCombinator(Promise<T> promise, size_t size = 0) : _size{size}, _promise{std::move(promise)} {
+  }
```
Comment (MBkkt) [inline include/yaclib/async/when_any.hpp:85]: Should be private

File: include/yaclib/async/when_any.hpp [hunk 8]
```diff
@@ -0,0 +1,162 @@
...
+    if (result) {
+      if (!_done.exchange(true, std::memory_order_acq_rel)) {
+        std::move(_promise).Set(std::move(result));
+      }
+    } else if (_size.fetch_sub(1, std::memory_order_acq_rel) <= 1U) {
+      std::move(_promise).Set(std::move(result));
+    }
+  }
+
+  explicit AnyCombinator(Promise<T> promise, size_t size = 0) : _size{size}, _promise{std::move(promise)} {
```
Comment (MBkkt) [inline include/yaclib/async/when_any.hpp:84]: Should be private

File: include/yaclib/async/when_any.hpp [hunk 9]
```diff
@@ -0,0 +1,162 @@
...
+
+  ~AnyCombinator() override {
+    if (!_done.load(std::memory_order_acquire)) {
+      std::move(_promise).Set(std::move(_except_error));
+    }
+  }
+
+ private:
+  alignas(kCacheLineSize) std::atomic<bool> _done{false};
+  alignas(kCacheLineSize) std::atomic<bool> _ex_er{false};
```
Comment (MBkkt) [inline include/yaclib/async/when_any.hpp:52]: ex_er is bad naming
Maybe _error?

File: include/yaclib/async/when_any.hpp [hunk 10]
```diff
@@ -0,0 +1,162 @@
...
+
+#include <yaclib/async/run.hpp>
+
+#include <iterator>
+#include <type_traits>
+#include <utility>
+
+namespace yaclib::async {
+
+enum class PolicyWhenAny { FirstError, LastError };
```
Comment (MBkkt) [inline include/yaclib/async/when_any.hpp:11]:
```suggestion
enum class PolicyWhenAny { FirstError, LastError, };
```

File: include/yaclib/async/when_all.hpp [hunk 11]
```diff
@@ -123,7 +125,7 @@ template <typename T, typename... Fs>
 auto WhenAll(Future<T>&& head, Fs&&... tail) {
   static_assert((... && util::IsFutureV<Fs>));  // TODO(kononovk): Add static assert that FutureValue<Fs> = T
   auto [future, combinator] = detail::AllCombinator<T, sizeof...(Fs) + 1>::Make();
-  detail::WhenAllImpl<sizeof...(Fs) + 1>(combinator, std::move(head), std::move(tail)...);
+  detail::WhenAllImpl<sizeof...(Fs) + 1>(combinator, std::move(head), std::forward<Fs>(tail)...);
```
Comment (MBkkt) [inline include/yaclib/async/when_all.hpp:128]: Plz add
c = nullptr

**Review by MBkkt — CHANGES_REQUESTED** [review]

File: include/yaclib/async/when_any.hpp [hunk 12]
```diff
@@ -0,0 +1,162 @@
...
+      if (!_done.exchange(true, std::memory_order_acq_rel)) {
+        std::move(_promise).Set(std::move(result));
+      }
+    } else if (!_ex_er.load(std::memory_order_acquire) && !_ex_er.exchange(true, std::memory_order_acq_rel)) {
+      _except_error = std::move(result);
+    }
+  }
+
+  explicit AnyCombinator(Promise<T> promise) : _promise{std::move(promise)} {
+  }
```
Comment (MBkkt) [inline include/yaclib/async/when_any.hpp:42]: Should be private

File: include/yaclib/async/when_any.hpp [hunk 13]
```diff
@@ -0,0 +1,162 @@
...
+  void Combine(util::Result<T>&& result) {
+    if (_done.load(std::memory_order_acquire)) {
+      return;
+    }
+
+    if (result) {
+      if (!_done.exchange(true, std::memory_order_acq_rel)) {
+        std::move(_promise).Set(std::move(result));
+      }
+    } else if (_size.fetch_sub(1, std::memory_order_acq_rel) <= 1U) {
```
Comment (MBkkt) [inline include/yaclib/async/when_any.hpp:79]: ==?

File: include/yaclib/async/when_any.hpp [hunk 14]
```diff
@@ -0,0 +1,162 @@
...
+  for (; begin != end; ++begin) {
+    std::move(*begin).Subscribe([c = combinator](util::Result<T>&& result) mutable {
+      c->Combine(std::move(result));
+      c = nullptr;
+    });
+  }
+  return std::move(future);
+}
+
+template <PolicyWhenAny P, typename T, typename... Fs>
```
Comment (MBkkt) [inline include/yaclib/async/when_any.hpp:148]: What about default policy?

**Review by MBkkt — APPROVED** [review]

**Review by MBkkt — APPROVED** [review]

**Pull request merged**